content/courses/advanced-ada/parts/resource_management/controlled_types.rst

```diff
@@ -1548,13 +1548,422 @@ Completion
+   :ref:`allocation check <Adv_Ada_Allocation_Check>`, which may raise a
+   :ada:`Program_Error` exception. In the code example for that section, we
+   used controlled types. Feel free to revisit the example.
+
 .. admonition:: Relevant topics
 
     - :arm22:`Completion and Finalization <7-6-1>`
 
+
+.. _Adv_Ada_Controlled_Types_Initialize_Exception:
+
+Exception raising in Initialize
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If an exception is raised in the :ada:`Initialize` procedure, we have abnormal
+completion. Let's see an example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Controlled_Types.Exception_Handling.CT_Initialize_Exception
+    :class: ada-run-expect-failure
+
+    with Ada.Finalization;
+
+    package CT_Initialize_Exception is
+
+       type T is tagged private;
+
+       procedure Dummy (E : T);
+
+    private
+
+       type T is new
+         Ada.Finalization.Controlled
+           with null record;
+
+       overriding
+       procedure Initialize (E : in out T);
+
+       overriding
+       procedure Finalize (E : in out T);
+
+    end CT_Initialize_Exception;
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    package body CT_Initialize_Exception is
+
+       function Int_Last return Integer is
+         (Integer'Last);
+
+       Cnt : Positive := Int_Last;
+
+       procedure Dummy (E : T) is
+       begin
+          Put_Line ("(Dummy...)");
+       end Dummy;
+
+       procedure Initialize (E : in out T) is
+       begin
+          Put_Line ("Initialize...");
+          Cnt := Cnt + 1;
+       end Initialize;
+
+       procedure Finalize (E : in out T) is
+       begin
+          Put_Line ("Finalize...");
+       end Finalize;
+
+    end CT_Initialize_Exception;
+
+    with CT_Initialize_Exception;
+    use  CT_Initialize_Exception;
+
+    procedure Show_Initialize_Exception is
+       A : T;
+    begin
+       Dummy (A);
+    end Show_Initialize_Exception;
+
+In the :ada:`Show_Initialize_Exception` procedure, we declare an object
+:ada:`A` of controlled type :ada:`T`. As we know, this declaration triggers a
+call to the :ada:`Initialize` procedure that we've implemented in the body of
+the :ada:`CT_Initialize_Exception` package. In the :ada:`Initialize` procedure,
+we're forcing an overflow to happen |mdash| by adding one to the :ada:`Cnt`
+variable, which already has the :ada:`Integer'Last` value.
+
+This is an example of abnormal completion, as the control is transferred out of
+the :ada:`Initialize` procedure, and the corresponding :ada:`Finalize`
+procedure is never called for object :ada:`A`.
+
+
+Bounded errors of controlled types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Bounded errors are an important topic when talking about exception and
+controlled types. In general, if an exception is raised in the :ada:`Adjust` or
+:ada:`Finalize` procedure, this is considered a bounded error |mdash| in its
+turn a bounded error raises a :ada:`Program_Error` exception.
```

Comment:
It raises the P_E exception only if the error is detected, I believe.

Reply:
Good point, I've changed the phrase accordingly.